### ISSUE TITLE
Valid JS code reported by TS as illegal code: Keywords cannot contain escape characters

### ISSUE THREAD

**Issue by jaitjacob** [microsoft/TypeScript#54196]
# Bug Report
### 🔎 Search Terms
\u{63}onstructor()
valid javascript code returns error
Keywords cannot contain escape characters
escape characters

### 🕗 Version & Regression Information

- This is the behavior in every version I tried.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=5.0.4#code/MYGwhgzhAEDC0G8BQ1oB0CuCBsBmAvgPYB2EALgE4bBmEUAUAlIvqgPRvQCmAjhgJYA3MCC7Ey0WgC5owEuSo06TFknxA)

### 💻 Code
```
class C {
  \u{63}onstructor() {}  // equivalent to: constructor() {}
}
```

### 🙁 Actual behavior
**Errors in code**
Keywords cannot contain escape characters.

### 🙂 Expected behavior

No errors as it is valid javascript code.

**Comment by fatcerberus** [comment]
Not related to the issue but why does JS even allow this?  I can't see any reason to ever do this unless you're trying to win some sort of code-obfuscation contest...

**Comment by guillaumebrunerie** [comment]
What makes you say that it is valid Javascript? As far as I can tell, the spec explicitly forbids escape sequences in keywords, which is consistent with the error message you get:

> NOTE 1
Per [5.1.5](https://tc39.es/ecma262/#sec-grammar-notation), keywords in the grammar match literal sequences of specific [SourceCharacter](https://tc39.es/ecma262/#prod-SourceCharacter) elements. A code point in a keyword cannot be expressed by a \ [UnicodeEscapeSequence](https://tc39.es/ecma262/#prod-UnicodeEscapeSequence).

(https://tc39.es/ecma262/#sec-identifier-names-static-semantics-early-errors)

**Comment by guillaumebrunerie** [comment]
Oh but it seems like `constructor` is not considered a keyword but merely a special identifier, in which case it should indeed be allowed I guess.

**Comment by Josh-Cena** [comment]
> Not related to the issue but why does JS even allow this? I can't see any reason to ever do this unless you're trying to win some sort of code-obfuscation contest...

It's more useful if you are trying to enter non-ASCII characters and you want to stay away from encoding issues.

**Comment by RyanCavanaugh** [comment]
I'd take a PR for this but if it even slightly complicates or slows down any other scenario, we're not going to merge it. If you really need to do this for whatever reason, you can put quotes around constructor and it'll parse fine again.

**Comment by fatcerberus** [comment]
@Josh-Cena Yeah, but stuff like `\u{63}onstructor` is just weird and I’m kind of surprised the spec goes out of its way to allow it.  I guess it makes sense if `constructor` is meant to be parsed as a normal identifier (where this preprocessing would just happen by default anyway) and the special behavior only applies afterwards…

**Comment by RyanCavanaugh** [comment]
In this parsing context, constructor is really more of an identifier than a keyword. That's why the quoted version works equivalently.

**Comment by Josh-Cena** [comment]
On a related note:

```ts
class C {
  constructor = 1;
}
```

I'm still unsure how TS distinguishes "grammar errors" from "diagnostics", but in this case it should be able to generate an AST and then report a diagnostic, instead of failing with "'(' expected".

**Comment by nmain** [comment]
> On a related note:

This is a syntax error in Javascript, although not due to parsing the syntax tree but due to an [early error](https://tc39.es/ecma262/#sec-class-definitions-static-semantics-early-errors).

**Comment by Josh-Cena** [comment]
That's right. Early errors are not exactly the same as the parser entering an invalid state, and it should still be able to provide an AST.
